File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/outtake/Rotate.java
package org.firstinspires.ftc.teamcode.outtake;

import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.teamcode.utils.ManualMechanismTeleOp;
import org.firstinspires.ftc.teamcode.utils.ServoPositionMechanism;

import static org.firstinspires.ftc.teamcode.Positions.Outtake.Rotate.*;

/**
 * Class representing the vertical rotation of the outtake claw.
 * It extends the ServoPositionMechanism class to control the rotation servo.
 */
public class Rotate extends ServoPositionMechanism {
    private final HardwareMap hardwareMap;
    /**
     * @param hardwareMap the hardware map to get the servo from.
     */
    public Rotate(HardwareMap hardwareMap) {
        super(transfer);
        this.hardwareMap = hardwareMap;
    }

    /**
     * Gets the servo associated with the Rotate mechanism.
     * @return an array containing the rotation servo.
     */
    @Override
    protected Servo[] getServos() {
        return new Servo[]{hardwareMap.get(Servo.class, "Outtake Rotate")};
    }
}

/**
 * TeleOp class for testing the Rotate mechanism manually.
 */
@TeleOp(name = "Outtake Rotate Test", group = "C Outtake")
class RotateTest extends ManualMechanismTeleOp {
    public RotateTest() {
        super(Rotate::new);
    }
}

/**
 * TeleOp class for testing the Rotate mechanism positions.
 * @noinspection DuplicatedCode
 */
@TeleOp(name = "Outtake Rotate Positions Test", group = "D Outtake")
class RotatePositions extends LinearOpMode {
    public static double current = 0.0;
    @Override
    public void runOpMode() {
        Rotate rotate = new Rotate(hardwareMap);
        waitForStart();
        while (opModeIsActive()) {
            if (gamepad1.dpad_down) current = transfer;
            else if (gamepad1.dpad_up) current = basket;
            else if (gamepad1.dpad_left) current = bar;
            else if (gamepad1.dpad_right) current = pickup;
            else current += rotate.getMultiplier() * (gamepad1.right_trigger - gamepad1.left_trigger);
            rotate.setTargetPosition(current);

            TelemetryPacket packet = new TelemetryPacket();
            if(!rotate.run(packet)) requestOpModeStop();
            telemetry.addData("Position", current);
            telemetry.update();
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/outtake/Rotate.kt
package org.firstinspires.ftc.outtake

import com.acmerobotics.dashboard.config.Config
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import com.qualcomm.robotcore.hardware.Servo
import org.firstinspires.ftc.outtake.Rotate.Companion.pickup
import org.firstinspires.ftc.outtake.Rotate.Companion.transfer
import org.firstinspires.ftc.outtake.Rotate.Companion.bar
import org.firstinspires.ftc.outtake.Rotate.Companion.basket
import org.firstinspires.ftc.utils.PressAction
import org.firstinspires.ftc.utils.ManualMechanismTeleOp
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.ServoPositionMechanism

@Config
class Rotate(hardwareMap: HardwareMap) : ServoPositionMechanism(transfer) {
    companion object {
        /**
         * The position of the outtake rotation when it is inside of the submersible.
         */
        @JvmField
        @Volatile
        var pickup = 0.834
        /**
         * The position of the outtake rotation when it is in the transfer position.
         */
        @JvmField
        @Volatile
        var transfer = 0.19
        /**
         * The position of the outtake rotation when it is in the high chamber position.
         */
        @JvmField
        @Volatile
        var bar = 0.71
        /**
         * The position of the outtake rotation when it is in the basket position.
         */
        @JvmField
        @Volatile
        var basket = 0.58
    }
    /**
     * Gets the servo associated with the Rotate mechanism.
     */
    override val servos: Array<Servo> = arrayOf(hardwareMap.servo["Outtake Rotate"])
}
/**
 * TeleOp class for testing the Rotate mechanism manually.
 */
@TeleOp(name = "Outtake Rotate Test", group = "C")
@Disabled
private class RotateTest : ManualMechanismTeleOp(::Rotate)
/**
 * TeleOp class for testing the positions of the Rotate mechanism.
 */
@TeleOp(name = "Outtake Rotate Position Test", group = "D")
@Disabled
private class RotatePositions : Movement() {
    /**
     * The Rotate mechanism to be tested.
     */
    override lateinit var system: Rotate

    override fun systemInit() {
        system = Rotate(hardwareMap)

        actions.add(PressAction(gamepad1::dpad_down) { system.targetPosition = pickup })
        actions.add(PressAction(gamepad1::dpad_up) { system.targetPosition = transfer })
        actions.add(PressAction(gamepad1::dpad_left) { system.targetPosition = basket })
        actions.add(PressAction(gamepad1::dpad_right) { system.targetPosition = bar })
    }

    override fun systemLoop() {
        // No need for a loop
    }
}